addons/binding/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/handler/NanoleafControllerHandler.java

```diff
@@ -0,0 +1,593 @@
+        boolean result = controllerListeners.add(controllerListener);
+        if (result) {
+            startPanelDiscoveryJob();
+        }
+        return result;
+    }
+
+    public boolean unregisterControllerListener(NanoleafControllerListener controllerListener) {
+        logger.debug("Unregister listener for controller {}", getThing().getUID());
+        boolean result = controllerListeners.remove(controllerListener);
+        if (result) {
+            stopPanelDiscoveryJob();
+        }
+        return result;
+    }
+
+    public NanoleafControllerConfig getControllerConfig() {
+        NanoleafControllerConfig config = new NanoleafControllerConfig();
+        config.address = this.getAddress();
+        config.port = this.getPort();
+        config.refreshInterval = this.getRefreshIntervall();
+        config.authToken = this.getAuthToken();
+        return config;
+    }
+
+    public synchronized void startPairingJob() {
+        if ((pairingJob == null || pairingJob.isCancelled())) {
+            logger.debug("Start pairing job, interval={} sec", PAIRING_INTERVAL);
+            pairingJob = scheduler.scheduleWithFixedDelay(this::runPairing, 0, PAIRING_INTERVAL, TimeUnit.SECONDS);
+        }
+    }
+
+    private synchronized void stopPairingJob() {
+        if (pairingJob != null && !pairingJob.isCancelled()) {
+            logger.debug("Stop pairing job");
+            pairingJob.cancel(true);
+            this.pairingJob = null;
+        }
+    }
+
+    private synchronized void startUpdateJob() {
+        if (StringUtils.isNotEmpty(getAuthToken())) {
+            if ((updateJob == null || updateJob.isCancelled())) {
+                logger.debug("Start controller status job, repeat every {} sec", getRefreshIntervall());
+                updateJob = scheduler.scheduleWithFixedDelay(this::runUpdate, 0, getRefreshIntervall(),
+                        TimeUnit.SECONDS);
+            }
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_PENDING,
+                    "@text/error.nanoleaf.controller.noToken");
+        }
+    }
+
+    private synchronized void stopUpdateJob() {
+        if (updateJob != null && !updateJob.isCancelled()) {
+            logger.debug("Stop status job");
+            updateJob.cancel(true);
+            this.updateJob = null;
+        }
+    }
+
+    public synchronized void startPanelDiscoveryJob() {
+        if (!controllerListeners.isEmpty() && (panelDiscoveryJob == null || panelDiscoveryJob.isCancelled())) {
+            logger.debug("Start panel discovery job, interval={} sec", PANEL_DISCOVERY_INTERVAL);
+            panelDiscoveryJob = scheduler.scheduleWithFixedDelay(this::runPanelDiscovery, 0, PANEL_DISCOVERY_INTERVAL,
+                    TimeUnit.SECONDS);
+        }
+    }
+
+    private synchronized void stopPanelDiscoveryJob() {
+        if (controllerListeners.isEmpty() && panelDiscoveryJob != null && !panelDiscoveryJob.isCancelled()) {

```

Comment:
The method `stopPanelDiscoveryJob` is called after the listener is removed from the list. So the job cancel might not be called because this empty check. It will probably just work because the job is never stopped, and when `startPanelDiscoveryJob`  is  called again no new job is called because the job is not null.
btw I think this start/stop is a good solution.

Reply:
Separate controller/bridge channels for hue, saturation and brightness have been removed. Same for the panel brightness channel. Color channels now also handle increase/decrease commands.

Reply:
I checked that `stopPanelDiscoveryJob` works correctly. Once the panel scan timeout is reached, the listener is removed, the `if` statement proofs true, and the panel discovery job gets cancelled.